ouroboros-network/src/Ouroboros/Network/Protocol/VersionNegotiation/Type.hs

```diff
@@ -0,0 +1,424 @@
+--
+decodeKnownVersions :: forall ver r s.
+                       ( Ord ver
+                       , Enum ver
+                       , Show ver
+                       , Serialise ver
+                       )
+                    => KnownVersions ver r
+                    -> CBOR.Decoder s (KnownVersions ver r)
+decodeKnownVersions knownVersions = CBOR.decodeMapLen >>= go Nothing []
+  where
+    go !ver0 !rs !l | l <= 0 = return $ KnownVersions $ reverse rs
+                    | otherwise = do
+      ver <- CBOR.decode
+      if Just ver <= ver0
+        then fail "decodeVersionMap: unsorted map"
+        else case findProtocolVersion ver knownVersions of
+          Nothing -> CBOR.decodeTerm >> go (Just ver) rs (pred l)
+          Just pv -> decodeVersionData pv >>= \r -> go (Just ver) (r : rs) (pred l)
+
+-- |
+-- Smart constructor for the sorted list of protocol versions.
+--
+newKnownVersions :: Ord ver => [ProtocolVersion ver r] -> KnownVersions ver r
+newKnownVersions ps = KnownVersions $ sortBy (\(ProtocolVersion verA _) (ProtocolVersion verB _) -> compare verA verB) ps
+
+-- |
+-- Find protocol version in a list of known versions.  It assumes that the
+-- @knownVersion@ list is sorted with respect to @ver@ (which is fullfiled by
+-- @newKnownVersions@ smart constructor).
+--
+findProtocolVersion
+  :: Ord ver
+  => ver
+  -> KnownVersions ver r
+  -> Maybe (ProtocolVersion ver r)
+findProtocolVersion ver (KnownVersions knownVersions) = go (length knownVersions) knownVersions
+    where
+      go _   []   = Nothing
+      go _   [pv] | protocolVersion pv == ver = Just pv
+                  | otherwise                 = Nothing
+      go len pvs  = let l = len `div` 2
+                    in case splitAt l pvs of
+                      (pvs', pv : pvs'') | protocolVersion pv == ver -> Just pv
+                                         | protocolVersion pv >  ver -> go l pvs'
+                                         | protocolVersion pv <  ver -> go (l - 1) pvs''
+                      _                  -> error "findProtocolVersion error"
+
+data ProtocolVersionDescription m where
+       ProtocolVersionDescription :: (Bounded ptcl, Enum ptcl)
+                                  => ptcl
+                                  -> MiniProtocolDescription ptcl m
+                                  -> ProtocolVersionDescription m
+
+-- |
+-- State transitions of the version negotation protocol.
+--
+data VersionNegotiationProtocol ver r where
+    StPropose :: VersionNegotiationProtocol ver r
+    StConfirm :: VersionNegotiationProtocol ver r
+    StDone    :: VersionNegotiationProtocol ver r
+
+
+instance Protocol (VersionNegotiationProtocol ver r) where
+
+    data Message (VersionNegotiationProtocol ver r) from to where
+
+      -- |
+      -- Propose a list of versions together with version data.  It must be
+      -- encoded to a sorted list.
+      --
+      MsgProposeVersions
+        :: KnownVersions ver r
```

Comment:
Suppose two peers have inconsistent `KnownVersions`: there's a version number for which the version data are of different types. Suppose as well that it's the highest version number, so it'll be picked.

When we run this protocol without a codec, there's no protocol error, but rather a failure case within the protocol (`VersionNegotiationTypeCastError`). On the other hand, when it's run *with* a codec, it's a decode failure, a violation of the protocol--unless the different version data types have compatible encodings, in which case the disagreement will probably arise as a later error, within other protocols determined by that version.

So the weird thing here is that the *codec* (and therefore the form of the wire protocol) is determined by those existential types in the `KnownVersions`, and I argue that's a bad choice. Instead, it should be determined by the type of the version number and some common type for the encoded version data (`ByteString` in practice, or `Dynamic`). This way we get coherence between running it purely using `connect`, and running it by way of a codec, which I believe is absolutely essential. Beyond that, it's just the right way forward from a design/semantics point of view: two peers that disagree on the *content* of the versions may still correctly run the version negotiation protocol itself without protocol error.

Reply:
Alex and I just discussed this. We think a sensible and clean approach might be to have the first stage (with the codec) return generic CBOR `Term`s, and then a second stage using `Term -> Maybe a` (for some appropriate `a`) for decoding the version info payloads. This makes the mini-protocol behaviour more independent of the codec, and makes the failure of decoding of the version info very clearly distinct from codec failure.

Reply:
So what @dcoutts proposes here can be translated to:
```
data VersionData ver verData r where
  VersionData
    :: ( Show verData
       , Eq verData
       , Typeable verData
       )
    => verData
    -> (verData -> verData -> r)
    -> VersionDataCodec verData
    -> VersionData ver verData r

data VersionDataCodec verData = VersionDataCodec {
    encodeVersionData :: verData -> Encoding,
    decodeVersionData :: CBor.Term -> Maybe verData
  }
```

We can further remove the `Show`, `Eq` and `Typeable` constraints using an explicit `Dict` (or `extra` like in type parameter which @avieth proposed). `Typeable` is not a problem since it's derived anyway, but at least `Show` might not be needed in some cases.